# Trabajo Practico 5 - Medidas Electrónicas 1
**Docentes:**

*   Ing. Marinsek, Emiliano
*   Ing. Perdomo, Juan Manuel

**Alumnos:**

*   Ferrario, Franco Ezequiel
*   Encinas, Leandro
*   Marchesi, Matias
*   Quiroga, Bruno

**Curso:** R4052

**Año:** 2026


# Librerias y funciones a utilizar

In [9]:
# Librerias de pyhon a utilizar
import numpy as np
import math
import IPython
from IPython.display import display, Math
import pandas as pd
from typing import Any


def incert_A(mediciones):
  s = np.std(mediciones, ddof=1, )  # ddof=1 -> grados de libertad v = 1/n-1
  return s/math.sqrt(len(mediciones))

#Incertidumbre tipo B para modo frecuencimetro
def incert_B(estabilidad_BT, delta_N, N, mediciones):
    """
    Calcula la incertidumbre Tipo B para el modo FRECUENCÍMETRO.
    """
    f_media = np.mean(mediciones)
    
    #Incertidumbre relativa del oscilador/base de tiempo
    u_bt_rel = estabilidad_BT / np.sqrt(3)
    
    #Incertidumbre relativa por el error de +/- 1 cuenta
    u_cuentas_rel = delta_N / (np.sqrt(3) * N)
    
    # Combinación de ambas componentes (relativas)
    return np.sqrt(u_bt_rel**2 + u_cuentas_rel**2)

# Adaptación de Incertidumbre Tipo B para el modo Periodímetro
def incert_B_periodimetro(estabilidad_BT, delta_N, N, resolucion_display):
    """
    Calcula la incertidumbre Tipo B combinando el error de la Base de Tiempo
    y el error de digitalización/resolución (distribución rectangular).
    """
    # Componente por estabilidad de la Base de Tiempo
    uB_bt = T_media * (estabilidad_BT / np.sqrt(3))
    
    # Componente por resolución del display (error de cuantización)
    # Se modela con distribución rectangular sobre el paso de visualización
    uB_res = resolucion_display / (2 * np.sqrt(3))
    
    # Combinación de ambas componentes Tipo B
    return np.sqrt(uB_bt**2 + uB_res**2)


def incert_B_ConMedia(err_rel, nd, ctas, media):
  return (err_rel/100 + nd/ctas) * media / math.sqrt(3)

def incert_Total(uA, uB):
  return math.sqrt(uA**2 + uB**2)

def incert_Comb_Rel(uX, uY,correlacion=0):
  return math.sqrt((uX['u_C']/uX['media'])**2 + (uY['u_C']/uY['media'])**2 - (2*(uX['u_C']/uX['media'])*(uY['u_C']/uY['media'])*correlacion))

# Generalidades de los instrumentos

El frecuencímetro utilizado para las prácticas es el **Protek U2000A**.

Para el cálculo de incertidumbres se utilizará la estabilidad de largo plazo de la
base de tiempo especificada por el fabricante, cuyo valor es:

$$
\frac{\Delta f_{BT}}{f_{BT}} = \pm 3\cdot10^{-7}/\text{mes}
$$

Como no se dispone de información sobre la antigüedad del instrumento, se
adopta como hipótesis una antigüedad de **5 años**, equivalentes a 60 meses.
Por lo tanto:

$$\frac{\Delta f_{BT}}{f_{BT}} = 3\cdot10^{-7}\cdot\frac{1}{\text{mes}}\cdot60\text{meses}$$

Expresando este valor en partes por millón:

$$
\frac{\Delta f_{BT}}{f_{BT}} = 18\,\text{ppm}
$$

y en porcentaje:

$$\frac{\Delta f_{BT}}{f_{BT}} = 18\cdot10^{-6} \cdot 100\% = 0.0018\%$$

Este valor será utilizado como error relativo de la base de tiempo en los
cálculos de incertidumbre.

A continuación, se coloca una celda con variables generales del instrumento que utilizaremos en todas las otras celdas.

In [5]:
k = 2

f_BT = 10e6                 # Frecuencia de la base de tiempo [Hz]

estabilidad_BT = 3e-7       # Estabilidad de largo plazo [1/mes]
antiguedad_meses = 5 * 12   # Antigüedad estimada del instrumento [meses]

error_rel_BT = estabilidad_BT * antiguedad_meses
error_rel_BT_pct = error_rel_BT * 100

nd = 1                      # Error de última cuenta/dígito

print(f"Base de tiempo: {f_BT/1e6} MHz")
print(f"Antigüedad estimada: {antiguedad_meses} meses")
print(f"Estabilidad acumulada: {error_rel_BT:.2e}")
print(f"Error relativo de BT: {error_rel_BT_pct:.4f} %")

Base de tiempo: 10.0 MHz
Antigüedad estimada: 60 meses
Estabilidad acumulada: 1.80e-05
Error relativo de BT: 0.0018 %


# Ejercicio 1

Las ecuaciones que se usarán para este ejercicio son las vistas en la cátedra

## Medición con frecuencia

Para la medición de frecuencia se usarán las ecuaciones:

$$f_I = 10^k \cdot f_{BT} \cdot N$$

Y su incertidumbre relativa:

$$\frac{\Delta f_I}{f_I} = \pm (\frac{\Delta f_{BT}}{f_{BT}} + \frac{\Delta N}{N})$$

Siendo el $\Delta N = 1$.

La cantidad de ciclos durante el gate time es:
$$N = f_x \cdot T_C$$.

Una vez obtenidos los datos se calcularán las incertidumbres.

## Medición con periodos promediados

Para la medición con periodos promediados se usarán las ecuaciones:
$$T_I = T_C \cdot N$$

Y su incertidumbre relativa:
$$\frac{\Delta T_I}{T_I} = \pm (\frac{\Delta f_{BT}}{f_{BT}} + \frac{\Delta N}{N} + \frac{e_D}{10^M})$$

Donde M es la cantidad de periodos promediados.
Además $e_D = 0$ porque estamos estudiando señales cuadradas, por lo que la incertdiumbre termina quedando:
$$\frac{\Delta T_I}{T_I} = \pm (\frac{\Delta f_{BT}}{f_{BT}} + \frac{\Delta N}{N})$$

## 10 Hz

Se medirá una señal cuadrada de $10Hz$ con $2V_{pp}$.

### 0.1s de Gate Time

Partiendo de las mediciones hechas en el laboratorio se harán las cuentas para 0.1s de Gate time.



In [10]:
# Datos 0.1s Gate Time
f_x = 10          # Hz
gate_time = 0.1   # s
delta_N = 1

mediciones = [
    0.01e3, 0.01e3, 0.01e3, 0.01e3, 0.01e3,
    0.01e3, 0.01e3, 0.01e3, 0.01e3, 0.01e3
]  # Hz

f_media = np.mean(mediciones)

# Cantidad de cuentas durante el gate time
N = f_media * gate_time

# Incertidumbre Tipo A
uA_f = incert_A(mediciones)

# Incertidumbre Tipo B

uB_f = incert_B(
    estabilidad_BT,
    delta_N,
    N,
    mediciones
)

# Incertidumbre total
uC_f = incert_Total(uA_f, uB_f)

U_f = k * uC_f

print(f"Media = {f_media:.2f} Hz")
print(f"uA = {uA_f:.4f} Hz")
print(f"uB = {uB_f:.4f} Hz")
print(f"uC = {uC_f:.4f} Hz")
print(f"U = {U_f:.4f} Hz")

Media = 10.00 Hz
uA = 0.0000 Hz
uB = 0.5774 Hz
uC = 0.5774 Hz
U = 1.1547 Hz


Finalmente se expresa el resultado:

<div style="text-align:center;"">
    <div style="border: 2px solid #555; padding: 5px; border-radius: 8px; text-align: center; background-color: #f9f9f9; display: inline-block;">
        $$f = (10.1 \pm 1.1)Hz$$
    </div>
</div>

### 1s de Gate Time

Partiendo de las mediciones hechas en el laboratorio se harán las cuentas para 1s de Gate time.

In [11]:
# Datos 1s Gate Time
f_x = 10          # Hz
gate_time = 1   # s
delta_N = 1

mediciones = [
    0.010e3, 0.010e3, 0.010e3, 0.010e3, 0.010e3,
    0.010e3, 0.010e3, 0.010e3, 0.010e3, 0.010e3
]  # Hz

f_media = np.mean(mediciones)

# Cantidad de cuentas durante el gate time
N = f_media * gate_time

# Incertidumbre Tipo A
uA_f = incert_A(mediciones)

# Incertidumbre Tipo B

uB_f = incert_B(
    estabilidad_BT,
    delta_N,
    N,
    mediciones
)

# Incertidumbre total
uC_f = incert_Total(uA_f, uB_f)

U_f = k * uC_f

print(f"Media = {f_media:.2f} Hz")
print(f"uA = {uA_f:.4f} Hz")
print(f"uB = {uB_f:.4f} Hz")
print(f"uC = {uC_f:.4f} Hz")
print(f"U = {U_f:.4f} Hz")

Media = 10.00 Hz
uA = 0.0000 Hz
uB = 0.0577 Hz
uC = 0.0577 Hz
U = 0.1155 Hz


Finalmente se expresa el resultado:

<div style="text-align:center;"">
    <div style="border: 2px solid #555; padding: 5px; border-radius: 8px; text-align: center; background-color: #f9f9f9; display: inline-block;">
        $$f = (10.0 \pm 1.2)Hz$$
    </div>
</div>

### 10s de Gate Time

Partiendo de las mediciones hechas en el laboratorio se harán las cuentas para 10s de Gate time.

In [12]:
# Datos 10s Gate Time
f_x = 10          # Hz
gate_time = 10   # s
delta_N = 1

mediciones = [
    0.0100e3, 0.0100e3, 0.0100e3, 0.0100e3, 0.0100e3,
    0.0100e3, 0.0100e3, 0.0100e3, 0.0100e3, 0.0100e3
]  # Hz

f_media = np.mean(mediciones)

# Cantidad de cuentas durante el gate time
N = f_media * gate_time

# Incertidumbre Tipo A
uA_f = incert_A(mediciones)

# Incertidumbre Tipo B

uB_f = incert_B(
    estabilidad_BT,
    delta_N,
    N,
    mediciones
)

# Incertidumbre total
uC_f = incert_Total(uA_f, uB_f)

U_f = k * uC_f

print(f"Media = {f_media:.2f} Hz")
print(f"uA = {uA_f:.4f} Hz")
print(f"uB = {uB_f:.4f} Hz")
print(f"uC = {uC_f:.4f} Hz")
print(f"U = {U_f:.4f} Hz")

Media = 10.00 Hz
uA = 0.0000 Hz
uB = 0.0058 Hz
uC = 0.0058 Hz
U = 0.0115 Hz


Finalmente se expresa el resultado:

<div style="text-align:center;"">
    <div style="border: 2px solid #555; padding: 5px; border-radius: 8px; text-align: center; background-color: #f9f9f9; display: inline-block;">
        $$f = (10.000 \pm 0.015)Hz$$
    </div>
</div>

### 1 Ciclo de periodos promediados

Partiendo de las mediciones hechas en el laboratorio se harán las cuentas para 1 ciclo de periodos promediados.

In [45]:
# ==========================================
# CONFIGURACIÓN DE DATOS (MODO PERIODÍMETRO)
# ==========================================

# Mediciones (1 µs expresado en segundos)
mediciones = [
    100003.0e-6,100002.9e-6,100003.0e-6,100002.9e-6,100003.0e-6,100002.9e-6,100003.0e-6,100002.9e-6,100003.0e-6,100002.9e-6

] # s


T_media = np.mean(mediciones)
M = 1 #Numero de ciclos

# En periodimetros, N representa las cuentas del reloj interno durante 1 ciclo de la señal
# N = T_media / T_clock = T_media * f_clock
N = M * T_media * f_BT

# Paso mínimo visible de tus datos actuales (para evaluar el display)
resolucion_display = 1e-6  # 1 µs

# CÁLCULO DE INCERTIDUMBRES
# Incertidumbre Tipo A
uA_T = incert_A(mediciones)

# Incertidumbre Tipo B (Adaptada a periodo)
uB_T = incert_B_periodimetro(estabilidad_BT, nd, N, resolucion_display)

# Incertidumbre total combinada
uC_T = incert_Total(uA_T, uB_T)

# Incertidumbre expandida
U_T = k * uC_T

# IMPRESIÓN DE RESULTADOS
print(f"Media = {T_media:.10e} s")
print(f"uA = {uA_T:.5e} s")
print(f"uB = {uB_T:.5e} s")
print(f"uC = {uC_T:.5e} s")
print(f"U  = {U_T:.5e} s")

Media = 1.0000295000e-01 s
uA = 1.66667e-08 s
uB = 2.89194e-07 s
uC = 2.89674e-07 s
U  = 5.79348e-07 s


Finalmente se expresa el resultado:

<div style="text-align:center;"">
    <div style="border: 2px solid #555; padding: 5px; border-radius: 8px; text-align: center; background-color: #f9f9f9; display: inline-block;">
        $$P = (100002.95 \pm 0.58) μs$$
    </div>
</div>

### 10 ciclo de periodos promediados
Partiendo de las mediciones hechas en el laboratorio se harán las cuentas para 10 ciclo de periodos promediados.

In [44]:
# ==========================================
# CONFIGURACIÓN DE DATOS (MODO PERIODÍMETRO)
# ==========================================

# Mediciones (1 µs expresado en segundos)
mediciones = [
100003.06e-6,
100003.09e-6,
100003.13e-6,
100003.16e-6,
100003.24e-6,
100003.28e-6,
100003.27e-6,
100003.29e-6,
100003.30e-6,
100003.29e-6
] # s

T_media = np.mean(mediciones)
M = 10 #Numero de ciclos
# En periodimetros, N representa las cuentas del reloj interno durante 1 ciclo de la señal
# N = T_media / T_clock = T_media * f_clock
N = M * T_media * f_BT

# Paso mínimo visible de tus datos actuales (para evaluar el display)
resolucion_display = 1e-6  # 1 µs

# CÁLCULO DE INCERTIDUMBRES
# Incertidumbre Tipo A
uA_T = incert_A(mediciones)

# Incertidumbre Tipo B (Adaptada a periodo)
uB_T = incert_B_periodimetro(estabilidad_BT, nd, N, resolucion_display)

# Incertidumbre total combinada
uC_T = incert_Total(uA_T, uB_T)

# Incertidumbre expandida
U_T = k * uC_T


# IMPRESIÓN DE RESULTADOS

print(f"Media = {T_media:.10e} s")
print(f"uA = {uA_T:.5e} s")
print(f"uB = {uB_T:.5e} s")
print(f"uC = {uC_T:.5e} s")
print(f"U  = {U_T:.5e} s")

Media = 1.0000321100e-01 s
uA = 2.90765e-08 s
uB = 2.89194e-07 s
uC = 2.90652e-07 s
U  = 5.81305e-07 s


Finalmente se expresa el resultado:

<div style="text-align:center;"">
    <div style="border: 2px solid #555; padding: 5px; border-radius: 8px; text-align: center; background-color: #f9f9f9; display: inline-block;">
        $$P = (100003.21 \pm 0.58) μs$$
    </div>
</div>

### 100 ciclo de periodos promediados
Partiendo de las mediciones hechas en el laboratorio se harán las cuentas para 100 ciclo de periodos promediados.

In [43]:
# ==========================================
# CONFIGURACIÓN DE DATOS (MODO PERIODÍMETRO)
# ==========================================

# Mediciones  (1 µs expresado en segundos)
mediciones = [
0
] # s

T_media = np.mean(mediciones)
M = 100 #Numero de ciclos
# En periodimetros, N representa las cuentas del reloj interno durante 1 ciclo de la señal
# N = T_media / T_clock = T_media * f_clock
N = M * T_media * f_BT

# Paso mínimo visible de tus datos actuales (para evaluar el display)
resolucion_display = 1e-6  # 1 µs

# CÁLCULO DE INCERTIDUMBRES
# Incertidumbre Tipo A
uA_T = incert_A(mediciones)

# Incertidumbre Tipo B (Adaptada a periodo)
uB_T = incert_B_periodimetro(estabilidad_BT, nd, N, resolucion_display)

# Incertidumbre total combinada
uC_T = incert_Total(uA_T, uB_T)

# Incertidumbre expandida
U_T = k * uC_T


# IMPRESIÓN DE RESULTADOS

print(f"Media = {T_media:.10e} s")
print(f"uA = {uA_T:.5e} s")
print(f"uB = {uB_T:.5e} s")
print(f"uC = {uC_T:.5e} s")
print(f"U  = {U_T:.5e} s")

Media = 0.0000000000e+00 s
uA = nan s
uB = 2.88675e-07 s
uC = nan s
U  = nan s


<div style="text-align:center;"">
    <div style="border: 2px solid #555; padding: 5px; border-radius: 8px; text-align: center; background-color: #f9f9f9; display: inline-block;">
        No pudimos realizar la medición ya que el instrumento falló por overflow
    </div>
</div>

## 1 KHz

### 0.1s de Gate Time

Partiendo de las mediciones hechas en el laboratorio se harán las cuentas para 0.1s de Gate time.



In [29]:
# Datos 0.1s Gate Time
f_x = 1e3          # Hz
gate_time = 0.1   # s
delta_N = 1

mediciones = [
    1e3,1e3,1e3,1e3,1e3,1e3,1e3,1e3,1e3,1e3
]  # Hz

f_media = np.mean(mediciones)

# Cantidad de cuentas durante el gate time
N = f_media * gate_time

# Incertidumbre Tipo A
uA_f = incert_A(mediciones)

# Incertidumbre Tipo B

uB_f = incert_B(
    estabilidad_BT,
    delta_N,
    N,
    mediciones
)

# Incertidumbre total
uC_f = incert_Total(uA_f, uB_f)

U_f = k * uC_f

print(f"Media = {f_media:.2f} Hz")
print(f"uA = {uA_f:.4f} Hz")
print(f"uB = {uB_f:.4f} Hz")
print(f"uC = {uC_f:.4f} Hz")
print(f"U = {U_f:.4f} Hz")

Media = 1000.00 Hz
uA = 0.0000 Hz
uB = 0.0058 Hz
uC = 0.0058 Hz
U = 0.0115 Hz


Finalmente se expresa el resultado:

<div style="text-align:center;"">
    <div style="border: 2px solid #555; padding: 5px; border-radius: 8px; text-align: center; background-color: #f9f9f9; display: inline-block;">
        $$f = (1000.000 \pm 0.011)Hz$$
    </div>
</div>

### 1s de Gate Time
Partiendo de las mediciones hechas en el laboratorio se harán las cuentas para 1s de Gate time.

In [32]:
# Datos 0.1s Gate Time
f_x = 1e3          # Hz
gate_time = 1   # s
delta_N = 1

mediciones = [
    1e3,1e3,1e3,1e3,1e3,1e3,1e3,1e3,1e3,1e3
]  # Hz

f_media = np.mean(mediciones)

# Cantidad de cuentas durante el gate time
N = f_media * gate_time

# Incertidumbre Tipo A
uA_f = incert_A(mediciones)

# Incertidumbre Tipo B

uB_f = incert_B(
    estabilidad_BT,
    delta_N,
    N,
    mediciones
)

# Incertidumbre total
uC_f = incert_Total(uA_f, uB_f)

U_f = k * uC_f

print(f"Media = {f_media:.4f} Hz")
print(f"uA = {uA_f:.4f} Hz")
print(f"uB = {uB_f:.4f} Hz")
print(f"uC = {uC_f:.4f} Hz")
print(f"U = {U_f:.4f} Hz")

Media = 1000.0000 Hz
uA = 0.0000 Hz
uB = 0.0006 Hz
uC = 0.0006 Hz
U = 0.0012 Hz


Finalmente se expresa el resultado:

<div style="text-align:center;"">
    <div style="border: 2px solid #555; padding: 5px; border-radius: 8px; text-align: center; background-color: #f9f9f9; display: inline-block;">
        $$f = (1000.0000 \pm 0.0012)Hz$$
    </div>
</div>

### 10s de Gate Time
Partiendo de las mediciones hechas en el laboratorio se harán las cuentas para 10s de Gate time.

In [35]:
# Datos 0.1s Gate Time
f_x = 1e3          # Hz
gate_time = 10   # s
delta_N = 1

mediciones = [
    1e3,1e3,1e3,1e3,1e3,1e3,1e3,1e3,1e3,1e3
]  # Hz

f_media = np.mean(mediciones)

# Cantidad de cuentas durante el gate time
N = f_media * gate_time

# Incertidumbre Tipo A
uA_f = incert_A(mediciones)

# Incertidumbre Tipo B

uB_f = incert_B(
    estabilidad_BT,
    delta_N,
    N,
    mediciones
)

# Incertidumbre total
uC_f = incert_Total(uA_f, uB_f)

U_f = k * uC_f

print(f"Media = {f_media:.5f} Hz")
print(f"uA = {uA_f:.5f} Hz")
print(f"uB = {uB_f:.5f} Hz")
print(f"uC = {uC_f:.5f} Hz")
print(f"U = {U_f:.5f} Hz")

Media = 1000.00000 Hz
uA = 0.00000 Hz
uB = 0.00006 Hz
uC = 0.00006 Hz
U = 0.00012 Hz


Finalmente se expresa el resultado:

<div style="text-align:center;"">
    <div style="border: 2px solid #555; padding: 5px; border-radius: 8px; text-align: center; background-color: #f9f9f9; display: inline-block;">
        $$f = (1000.00000 \pm 0.00012)Hz$$
    </div>
</div>

### 1 ciclo de periodos promediados
Partiendo de las mediciones hechas en el laboratorio se harán las cuentas para 1 ciclo de periodos promediados.

In [42]:
# ==========================================
# CONFIGURACIÓN DE DATOS (MODO PERIODÍMETRO)
# ==========================================

# Mediciones(1 µs expresado en segundos)
mediciones = [
   1000.0e-6,1000.1e-6,1000.0e-6,1000.1e-6,1000.0e-6,1000.1e-6,1000.0e-6,1000.1e-6,1000.0e-6,1000.1e-6
] # s


T_media = np.mean(mediciones)
M = 1 #Numero de ciclos

# En periodimetros, N representa las cuentas del reloj interno durante 1 ciclo de la señal
# N = T_media / T_clock = T_media * f_clock
N = M * T_media * f_BT

# Paso mínimo visible de tus datos actuales (para evaluar el display)
resolucion_display = 1e-6  # 1 µs

# CÁLCULO DE INCERTIDUMBRES
# Incertidumbre Tipo A
uA_T = incert_A(mediciones)

# Incertidumbre Tipo B (Adaptada a periodo)
uB_T = incert_B_periodimetro(estabilidad_BT, nd, N, resolucion_display)

# Incertidumbre total combinada
uC_T = incert_Total(uA_T, uB_T)

# Incertidumbre expandida
U_T = k * uC_T

# IMPRESIÓN DE RESULTADOS
print(f"Media = {T_media:.10e} s")
print(f"uA = {uA_T:.5e} s")
print(f"uB = {uB_T:.5e} s")
print(f"uC = {uC_T:.5e} s")
print(f"U  = {U_T:.5e} s")

Media = 1.0000500000e-03 s
uA = 1.66667e-08 s
uB = 2.88675e-07 s
uC = 2.89156e-07 s
U  = 5.78312e-07 s


Finalmente se expresa el resultado:

<div style="text-align:center;"">
    <div style="border: 2px solid #555; padding: 5px; border-radius: 8px; text-align: center; background-color: #f9f9f9; display: inline-block;">
        $$P = (1000.05 \pm 0.58) μs$$
    </div>
</div>

### 10 ciclo de periodos promediados
Partiendo de las mediciones hechas en el laboratorio se harán las cuentas para 10 ciclo de periodos promediados.

In [46]:
# ==========================================
# CONFIGURACIÓN DE DATOS (MODO PERIODÍMETRO)
# ==========================================

# Mediciones(1 µs expresado en segundos)
mediciones = [
1000.03e-6,
1000.04e-6,
1000.03e-6,
1000.04e-6,
1000.03e-6,
1000.04e-6,
1000.03e-6,
1000.04e-6,
1000.03e-6,
1000.04e-6
] # s


T_media = np.mean(mediciones)
M = 10 #Numero de ciclos

# En periodimetros, N representa las cuentas del reloj interno durante 1 ciclo de la señal
# N = T_media / T_clock = T_media * f_clock
N = M * T_media * f_BT

# Paso mínimo visible de tus datos actuales (para evaluar el display)
resolucion_display = 1e-6  # 1 µs

# CÁLCULO DE INCERTIDUMBRES
# Incertidumbre Tipo A
uA_T = incert_A(mediciones)

# Incertidumbre Tipo B (Adaptada a periodo)
uB_T = incert_B_periodimetro(estabilidad_BT, nd, N, resolucion_display)

# Incertidumbre total combinada
uC_T = incert_Total(uA_T, uB_T)

# Incertidumbre expandida
U_T = k * uC_T

# IMPRESIÓN DE RESULTADOS
print(f"Media = {T_media:.10e} s")
print(f"uA = {uA_T:.5e} s")
print(f"uB = {uB_T:.5e} s")
print(f"uC = {uC_T:.5e} s")
print(f"U  = {U_T:.5e} s")

Media = 1.0000350000e-03 s
uA = 1.66667e-09 s
uB = 2.88675e-07 s
uC = 2.88680e-07 s
U  = 5.77360e-07 s


Finalmente se expresa el resultado:

<div style="text-align:center;"">
    <div style="border: 2px solid #555; padding: 5px; border-radius: 8px; text-align: center; background-color: #f9f9f9; display: inline-block;">
        $$P = (1000.03 \pm 0.58) μs$$
    </div>
</div>

### 100 ciclo de periodos promediados
Partiendo de las mediciones hechas en el laboratorio se harán las cuentas para 100 ciclo de periodos promediados.

In [47]:
# ==========================================
# CONFIGURACIÓN DE DATOS (MODO PERIODÍMETRO)
# ==========================================

# Mediciones(1 µs expresado en segundos)
mediciones = [
1000.035e-6,
1000.036e-6,
1000.035e-6,
1000.036e-6,
1000.035e-6,
1000.036e-6,
1000.035e-6,
1000.036e-6,
1000.035e-6,
1000.036e-6
] # s


T_media = np.mean(mediciones)
M = 100 #Numero de ciclos

# En periodimetros, N representa las cuentas del reloj interno durante 1 ciclo de la señal
# N = T_media / T_clock = T_media * f_clock
N = M * T_media * f_BT

# Paso mínimo visible de tus datos actuales (para evaluar el display)
resolucion_display = 1e-6  # 1 µs

# CÁLCULO DE INCERTIDUMBRES
# Incertidumbre Tipo A
uA_T = incert_A(mediciones)

# Incertidumbre Tipo B (Adaptada a periodo)
uB_T = incert_B_periodimetro(estabilidad_BT, nd, N, resolucion_display)

# Incertidumbre total combinada
uC_T = incert_Total(uA_T, uB_T)

# Incertidumbre expandida
U_T = k * uC_T

# IMPRESIÓN DE RESULTADOS
print(f"Media = {T_media:.10e} s")
print(f"uA = {uA_T:.5e} s")
print(f"uB = {uB_T:.5e} s")
print(f"uC = {uC_T:.5e} s")
print(f"U  = {U_T:.5e} s")

Media = 1.0000355000e-03 s
uA = 1.66667e-10 s
uB = 2.88675e-07 s
uC = 2.88675e-07 s
U  = 5.77350e-07 s


Finalmente se expresa el resultado:

<div style="text-align:center;"">
    <div style="border: 2px solid #555; padding: 5px; border-radius: 8px; text-align: center; background-color: #f9f9f9; display: inline-block;">
        $$P = (1000.03 \pm 0.58) μs$$
    </div>
</div>

### 1MHz

### 0.1s de Gate Time

Partiendo de las mediciones hechas en el laboratorio se harán las cuentas para 0.1s de Gate time.



In [49]:
# Datos 0.1s Gate Time
f_x = 1e6          # Hz
gate_time = 0.1   # s
delta_N = 1

mediciones = [
    999.96e3,999.97e3,999.96e3,999.97e3,999.96e3,999.97e3,999.96e3,999.97e3,999.96e3,999.97e3
]  # Hz

f_media = np.mean(mediciones)

# Cantidad de cuentas durante el gate time
N = f_media * gate_time

# Incertidumbre Tipo A
uA_f = incert_A(mediciones)

# Incertidumbre Tipo B

uB_f = incert_B(
    estabilidad_BT,
    delta_N,
    N,
    mediciones
)

# Incertidumbre total
uC_f = incert_Total(uA_f, uB_f)

U_f = k * uC_f

print(f"Media = {f_media:.4f} Hz")
print(f"uA = {uA_f:.4f} Hz")
print(f"uB = {uB_f:.4f} Hz")
print(f"uC = {uC_f:.4f} Hz")
print(f"U = {U_f:.4f} Hz")


Media = 999965.0000 Hz
uA = 1.6667 Hz
uB = 0.0000 Hz
uC = 1.6667 Hz
U = 3.3333 Hz


Finalmente se expresa el resultado:

<div style="text-align:center;"">
    <div style="border: 2px solid #555; padding: 5px; border-radius: 8px; text-align: center; background-color: #f9f9f9; display: inline-block;">
        $$f = (999965.0 \pm 3.3)Hz$$
    </div>
</div>

### 1s de Gate Time

Partiendo de las mediciones hechas en el laboratorio se harán las cuentas para 1s de Gate time.



In [53]:
# Datos 1s Gate Time
f_x = 1e6          # Hz
gate_time = 1   # s
delta_N = 1

mediciones = [
999.965e3,
999.964e3,
999.964e3,
999.965e3,
999.965e3,
999.965e3,
999.965e3,
999.965e3,
999.964e3,
999.964e3
]  # Hz

f_media = np.mean(mediciones)

# Cantidad de cuentas durante el gate time
N = f_media * gate_time

# Incertidumbre Tipo A
uA_f = incert_A(mediciones)

# Incertidumbre Tipo B

uB_f = incert_B(
    estabilidad_BT,
    delta_N,
    N,
    mediciones
)

# Incertidumbre total
uC_f = incert_Total(uA_f, uB_f)

U_f = k * uC_f

print(f"Media = {f_media:.4f} Hz")
print(f"uA = {uA_f:.4f} Hz")
print(f"uB = {uB_f:.4f} Hz")
print(f"uC = {uC_f:.4f} Hz")
print(f"U = {U_f:.4f} Hz")


Media = 999964.6000 Hz
uA = 0.1633 Hz
uB = 0.0000 Hz
uC = 0.1633 Hz
U = 0.3266 Hz


Finalmente se expresa el resultado:

<div style="text-align:center;"">
    <div style="border: 2px solid #555; padding: 5px; border-radius: 8px; text-align: center; background-color: #f9f9f9; display: inline-block;">
        $$f = (999964.60 \pm 0.33)Hz$$
    </div>
</div>

### 10s de Gate Time

Partiendo de las mediciones hechas en el laboratorio se harán las cuentas para 10s de Gate time.



In [55]:
# Datos 1s Gate Time
f_x = 1e6          # Hz
gate_time = 10   # s
delta_N = 1

mediciones = [
999.9648e3,
999.9648e3,
999.9648e3,
999.9647e3,
999.9648e3,
999.9648e3,
999.9648e3,
999.9647e3,
999.9648e3,
999.9648e3
]  # Hz

f_media = np.mean(mediciones)

# Cantidad de cuentas durante el gate time
N = f_media * gate_time

# Incertidumbre Tipo A
uA_f = incert_A(mediciones)

# Incertidumbre Tipo B

uB_f = incert_B(
    estabilidad_BT,
    delta_N,
    N,
    mediciones
)

# Incertidumbre total
uC_f = incert_Total(uA_f, uB_f)

U_f = k * uC_f

print(f"Media = {f_media:.4f} Hz")
print(f"uA = {uA_f:.4f} Hz")
print(f"uB = {uB_f:.4f} Hz")
print(f"uC = {uC_f:.4f} Hz")
print(f"U = {U_f:.4f} Hz")


Media = 999964.7800 Hz
uA = 0.0133 Hz
uB = 0.0000 Hz
uC = 0.0133 Hz
U = 0.0267 Hz


Finalmente se expresa el resultado:

<div style="text-align:center;"">
    <div style="border: 2px solid #555; padding: 5px; border-radius: 8px; text-align: center; background-color: #f9f9f9; display: inline-block;">
        $$f = (999964.780 \pm 0.027)Hz$$
    </div>
</div>

### 1 ciclo de periodos promediados
Partiendo de las mediciones hechas en el laboratorio se harán las cuentas para 1 ciclo de periodos promediados.

In [58]:
# ==========================================
# CONFIGURACIÓN DE DATOS (MODO PERIODÍMETRO)
# ==========================================

# Mediciones(1 µs expresado en segundos)
mediciones = [
1e-6,
1e-6,
1e-6,
1e-6,
1e-6,
1e-6,
1e-6,
1e-6,
1e-6,
1e-6
] # s


T_media = np.mean(mediciones)
M = 1 #Numero de ciclos

# En periodimetros, N representa las cuentas del reloj interno durante 1 ciclo de la señal
# N = T_media / T_clock = T_media * f_clock
N = M * T_media * f_BT

# Paso mínimo visible de tus datos actuales (para evaluar el display)
resolucion_display = 1e-6  # 1 µs

# CÁLCULO DE INCERTIDUMBRES
# Incertidumbre Tipo A
uA_T = incert_A(mediciones)

# Incertidumbre Tipo B (Adaptada a periodo)
uB_T = incert_B_periodimetro(estabilidad_BT, nd, N, resolucion_display)

# Incertidumbre total combinada
uC_T = incert_Total(uA_T, uB_T)

# Incertidumbre expandida
U_T = k * uC_T

# IMPRESIÓN DE RESULTADOS
print(f"Media = {T_media:.10e} s")
print(f"uA = {uA_T:.5e} s")
print(f"uB = {uB_T:.5e} s")
print(f"uC = {uC_T:.5e} s")
print(f"U  = {U_T:.5e} s")

Media = 1.0000000000e-06 s
uA = 7.05861e-23 s
uB = 2.88675e-07 s
uC = 2.88675e-07 s
U  = 5.77350e-07 s


Finalmente se expresa el resultado:

<div style="text-align:center;"">
    <div style="border: 2px solid #555; padding: 5px; border-radius: 8px; text-align: center; background-color: #f9f9f9; display: inline-block;">
        $$P = (1000.00 \pm 0.57) μs$$
    </div>
</div>

### 10 ciclo de periodos promediados
Partiendo de las mediciones hechas en el laboratorio se harán las cuentas para 10 ciclo de periodos promediados.

In [60]:
# ==========================================
# CONFIGURACIÓN DE DATOS (MODO PERIODÍMETRO)
# ==========================================

# Mediciones(1 µs expresado en segundos)
mediciones = [
1e-6,
1e-6,
1e-6,
1e-6,
1e-6,
1e-6,
1e-6,
1e-6,
1e-6,
1e-6
] # s


T_media = np.mean(mediciones)
M = 10 #Numero de ciclos

# En periodimetros, N representa las cuentas del reloj interno durante 1 ciclo de la señal
# N = T_media / T_clock = T_media * f_clock
N = M * T_media * f_BT

# Paso mínimo visible de tus datos actuales (para evaluar el display)
resolucion_display = 1e-6  # 1 µs

# CÁLCULO DE INCERTIDUMBRES
# Incertidumbre Tipo A
uA_T = incert_A(mediciones)

# Incertidumbre Tipo B (Adaptada a periodo)
uB_T = incert_B_periodimetro(estabilidad_BT, nd, N, resolucion_display)

# Incertidumbre total combinada
uC_T = incert_Total(uA_T, uB_T)

# Incertidumbre expandida
U_T = k * uC_T

# IMPRESIÓN DE RESULTADOS
print(f"Media = {T_media:.10e} s")
print(f"uA = {uA_T:.5e} s")
print(f"uB = {uB_T:.5e} s")
print(f"uC = {uC_T:.5e} s")
print(f"U  = {U_T:.5e} s")

Media = 1.0000000000e-06 s
uA = 7.05861e-23 s
uB = 2.88675e-07 s
uC = 2.88675e-07 s
U  = 5.77350e-07 s


Finalmente se expresa el resultado:

<div style="text-align:center;"">
    <div style="border: 2px solid #555; padding: 5px; border-radius: 8px; text-align: center; background-color: #f9f9f9; display: inline-block;">
        $$P = (1000.00 \pm 0.58) μs$$
    </div>
</div>

### 100 ciclo de periodos promediados
Partiendo de las mediciones hechas en el laboratorio se harán las cuentas para 100 ciclo de periodos promediados.

In [62]:
# ==========================================
# CONFIGURACIÓN DE DATOS (MODO PERIODÍMETRO)
# ==========================================

# Mediciones(1 µs expresado en segundos)
mediciones = [
1e-6,
1e-6,
1e-6,
1e-6,
1e-6,
1e-6,
1e-6,
1e-6,
1e-6,
1e-6
] # s


T_media = np.mean(mediciones)
M = 100 #Numero de ciclos

# En periodimetros, N representa las cuentas del reloj interno durante 1 ciclo de la señal
# N = T_media / T_clock = T_media * f_clock
N = M * T_media * f_BT

# Paso mínimo visible de tus datos actuales (para evaluar el display)
resolucion_display = 1e-6  # 1 µs

# CÁLCULO DE INCERTIDUMBRES
# Incertidumbre Tipo A
uA_T = incert_A(mediciones)

# Incertidumbre Tipo B (Adaptada a periodo)
uB_T = incert_B_periodimetro(estabilidad_BT, nd, N, resolucion_display)

# Incertidumbre total combinada
uC_T = incert_Total(uA_T, uB_T)

# Incertidumbre expandida
U_T = k * uC_T

# IMPRESIÓN DE RESULTADOS
print(f"Media = {T_media:.10e} s")
print(f"uA = {uA_T:.5e} s")w
print(f"uB = {uB_T:.5e} s")
print(f"uC = {uC_T:.5e} s")
print(f"U  = {U_T:.5e} s")

Media = 1.0000000000e-06 s
uA = 7.05861e-23 s
uB = 2.88675e-07 s
uC = 2.88675e-07 s
U  = 5.77350e-07 s


Finalmente se expresa el resultado:

<div style="text-align:center;"">
    <div style="border: 2px solid #555; padding: 5px; border-radius: 8px; text-align: center; background-color: #f9f9f9; display: inline-block;">
        $$P = (1000.00 \pm 0.58) μs$$
    </div>
</div>

# Ejercicio 2

El ciclo de actividad y el periodo de una serie de señales rectangulares de 1kHz con un 20%, 50% y 80% de dicho ciclo utilizando el modo de medición de intervalos (TI A→B).

### Definición de Intervalos de Tiempo (TI A$\rightarrow$B)

A través del modo de medición de intervalos de tiempo entre los canales A y B, se determinan los parámetros geométricos de la señal rectangular según la combinación de flancos configurada:

* **$(+, +)$**: Mide el intervalo desde un flanco ascendente hasta el siguiente flanco ascendente. Representa el **Periodo ($T$)**.
* **$(-, -)$**: Mide el intervalo desde un flanco descendente hasta el siguiente flanco descendente. Representa equivalentemente el **Periodo ($T$)**.
* **$(+, -)$**: Mide el intervalo desde un flanco ascendente hasta el flanco descendente inmediato. Representa el **Ancho de Pulso Positivo ($\tau_+$)** (tiempo en estado alto).
* **$(-, +)$**: Mide el intervalo desde un flanco descendente hasta el flanco ascendente inmediato. Representa el **Ancho de Pulso Negativo ($\tau_-$)** (tiempo en estado bajo).

### Cálculo del Ciclo de Actividad (*Duty Cycle*)

A partir de estas combinaciones, el ciclo de actividad o *Duty Cycle* ($D$) se calcula formalmente mediante la relación entre el ancho de pulso positivo y el periodo total de la señal:

$$D = \frac{\tau_+}{T}$$

Donde $\tau_+$ es el valor promedio obtenido mediante la configuración de flancos $(+, -)$ y $T$ es el periodo promedio obtenido mediante la configuración $(+, +)$.



### Propagación de Incertidumbre del Ciclo de Actividad ($D$)

Dado que el *Duty Cycle* es una magnitud indirecta que depende de dos variables medidas de forma independiente ($\tau_+$ y $T$), su incertidumbre combinada estándar $u_c(D)$ se calcula aplicando la ley de propagación de la incertidumbre mediante el método de las derivadas parciales (Guía GUM), asumiendo una covarianza despreciable:

$$u_c(D) = \sqrt{\left( \frac{\partial D}{\partial \tau_+} \cdot u_c(\tau_+) \right)^2 + \left( \frac{\partial D}{\partial T} \cdot u_c(T) \right)^2}$$

#### 1. Coeficientes de Sensibilidad

Los coeficientes de sensibilidad representan el impacto que tiene la variación de cada variable medida sobre el resultado final. Se calculan derivando parcialmente la función $D = \frac{\tau_+}{T}$:

*   **Respecto al ancho de pulso ($c_{\tau_+}$):**
    $$c_{\tau_+} = \frac{\partial D}{\partial \tau_+} = \frac{1}{T}$$

*   **Respecto al periodo ($c_T$):**
    $$c_T = \frac{\partial D}{\partial T} = -\frac{\tau_+}{T^2}$$

#### 2. Expresión Final de la Incertidumbre Combinada

Sustituyendo los coeficientes de sensibilidad en la ecuación original, obtenemos la fórmula analítica que está programada en nuestro algoritmo de Python:

$$u_c(D) = \sqrt{\left( \frac{1}{T} \cdot u_c(\tau_+) \right)^2 + \left( -\frac{\tau_+}{T^2} \cdot u_c(T) \right)^2}$$

Finalmente, la **incertidumbre expandida** para un nivel de confianza de aproximadamente el 95% utilizando un factor de cobertura $k=2$ se expresa como:

$$U(D) = k \cdot u_c(D)$$

In [11]:
import numpy as np

# ==========================================
# FUNCIONES DE EVALUACIÓN METROLÓGICA
# ==========================================

def calcular_uB_intervalo(valor_medio, estabilidad_BT, resolucion=1e-7):
    """
    Calcula la incertidumbre Tipo B para intervalos de tiempo (TI A->B).
    Incluye el error de la Base de Tiempo y la resolución de cuantización.
    """
    uB_bt = valor_medio * (estabilidad_BT / np.sqrt(3))
    uB_res = resolucion / (2 * np.sqrt(3))
    return np.sqrt(uB_bt**2 + uB_res**2)

# ==========================================
# INGRESO DE DATOS (REEMPLAZÁ CON TUS MUESTRAS)
# ==========================================

# Parámetros instrumentales del U2000A
estabilidad_BT = 5.3e-6  
resolucion_ti = 1e-7     # 100 ns típicamente para intervalos de tiempo en este modo
k = 2                    # Factor de cobertura

# Estructura de diccionario para almacenar las 10 muestras de cada combinación.
# Reemplazá las listas vacías [] con tus 10 mediciones en segundos (ej. 200e-6).
datos_experimento = {
    "20%": {
        "++": [0.001000066, 0.001000069, 0.001000066, 0.001000069, 0.001000066, 0.001000069, 0.001000066, 0.001000069, 0.001000066, 0.001000069], # Periodo T
        "+-": [0.000199274, 0.000199275, 0.000199283, 0.00019928, 0.000199279, 0.000199274, 0.000199275, 0.000199283, 0.00019928, 0.000199279], # Pulso positivo tau
        "-+": [0.000800816, 0.000800818, 0.00080082, 0.000800823, 0.000800816, 0.000800818, 0.00080082, 0.000800823, 0.000800821, 0.000800822], # Pulso negativo
        "--": [0.001000068, 0.001000069, 0.001000068, 0.001000069, 0.001000068, 0.001000069, 0.001000068, 0.001000069, 0.001000066, 0.001000069]  # Periodo alternativo T_down
    },
    "50%": {
        "++": [0.001000066, 0.001000069, 0.001000066, 0.001000069, 0.001000066, 0.001000069, 0.001000066, 0.001000069, 0.001000066, 0.001000069],
        "+-": [0.000500052, 0.000500062, 0.000500052, 0.000500062, 0.000500052, 0.000500062, 0.000500052, 0.000500062, 0.000500052, 0.000500062],
        "-+": [0.000500041, 0.000500039, 0.000500053, 0.000500045, 0.000500041, 0.000500039, 0.000500053, 0.000500045, 0.000500039, 0.000500039],
        "--": [0.001000066, 0.001000069, 0.001000066, 0.001000069, 0.001000066, 0.001000069, 0.001000066, 0.001000069, 0.001000066, 0.001000069]
    },
    "80%": {
        "++": [0.001000062, 0.001000064, 0.001000062, 0.001000064, 0.001000062, 0.001000064, 0.001000062, 0.001000064, 0.001000062, 0.001000064],
        "+-": [0.000799868, 0.000799864, 0.000799869, 0.000799868, 0.000799864, 0.000799869, 0.000799868, 0.000799864, 0.000799869, 0.000799864],
        "-+": [0.000200242, 0.00020024, 0.000200241, 0.000200242, 0.00020024, 0.000200241, 0.000200242, 0.00020024, 0.000200241, 0.00020024],
        "--": [0.001000068, 0.001000069, 0.001000068, 0.001000069, 0.001000068, 0.001000069, 0.001000068, 0.001000069, 0.001000068, 0.001000069]
    }
}

# ==========================================
# PROCESAMIENTO METROLÓGICO
# ==========================================

print(f"{'Duty Objetivo':<15} | {'Periodo T (s)':<15} | {'U_T (s)':<12} | {'Duty Calc (%)':<15} | {'U_Duty (%)':<12}")
print("-" * 80)

for duty, flancos in datos_experimento.items():
    # Validamos que existan datos cargados en las combinaciones principales (++ y +-)
    if len(flancos["++"]) == 0 or len(flancos["+-"]) == 0:
        print(f"{duty:<15} | [Sin datos suficientes en ++ o +-]")
        continue
        
    # 1. Procesamiento del Periodo T (Flanco ++)
    muestras_T = np.array(flancos["++"])
    T_media = np.mean(muestras_T)
    uA_T = incert_A(muestras_T)
    uB_T = calcular_uB_intervalo(T_media, estabilidad_BT, resolucion_ti)
    uc_T = np.sqrt(uA_T**2 + uB_T**2)
    U_T = k * uc_T
    
    # 2. Procesamiento del Ancho de Pulso tau (Flanco +-)
    muestras_tau = np.array(flancos["+-"])
    tau_media = np.mean(muestras_tau)
    uA_tau = calcular_uA(muestras_tau)
    uB_tau = calcular_uB_intervalo(tau_media, estabilidad_BT, resolucion_ti)
    uc_tau = np.sqrt(uA_tau**2 + uB_tau**2)
    
    # 3. Cálculo del Duty Cycle y Propagación de Incertidumbre (GUM)
    # D = tau / T
    D_media = tau_media / T_media
    
    # Coeficientes de sensibilidad (derivadas parciales)
    c_tau = 1 / T_media
    c_T = -tau_media / (T_media**2)
    
    # Incertidumbre combinada de D (asumiendo covarianza despreciable)
    uc_D = np.sqrt((c_tau * uc_tau)**2 + (c_T * uc_T)**2)
    U_D = k * uc_D
    
    # Impresión formateada con notación científica y porcentajes
    print(f"{duty:<15} | {T_media:.6e} | {U_T:.2e} | {D_media*100:.3f}% | {U_D*100:.2e}%")


Duty Objetivo   | Periodo T (s)   | U_T (s)      | Duty Calc (%)   | U_Duty (%)  
--------------------------------------------------------------------------------
20%             | 1.000068e-03 | 5.81e-08 | 19.926% | 5.89e-03%
50%             | 1.000068e-03 | 5.81e-08 | 50.002% | 6.48e-03%
80%             | 1.000063e-03 | 5.81e-08 | 79.982% | 7.43e-03%


### Resultados

Los valores medidos y sus incertidumbres expandidas asociadas:

| Parámetro / Condición | Periodo $T$ (s) | Duty Calc (%) |
| :--- | :---: | :---: |
| **Duty Objetivo: 20%** | $(1.000068 \pm 0.000058) \times 10^{-3}$ | $19.9260 \pm 0.0059$ |
| **Duty Objetivo: 50%** | $(1.000068 \pm 0.000058) \times 10^{-3}$ | $50.0020 \pm 0.0065$ |
| **Duty Objetivo: 80%** | $(1.000063 \pm 0.000058) \times 10^{-3}$ | $79.9820 \pm 0.0074$ |

*Las incertidumbres reportadas corresponden a la incertidumbre expandida del sistema, obtenidas a partir de una incertidumbre combinada estándar multiplicada por un factor de cobertura $k = 2$,  un intervalo con un nivel de confianza de aproximadamente el 95%.*


# Ejercicio 3


La relación de frecuencia entre una señal senoidal de 10MHz con una de 1MHz utilizando el modo de medición de relación de frecuencias (Ratio A/B) con la mayor resolución posible.


### Medición de Relación de Frecuencias (Ratio $A/B$)

El modo de relación de frecuencias determina cuántas veces entra la frecuencia de una señal conectada al Canal A ($f_A$) respecto a una señal de referencia conectada al Canal B ($f_B$). Para lograr la **mayor resolución posible**, el instrumento maximiza el tiempo de compuerta (*Gate Time*), promediando las cuentas sobre un número elevado de ciclos de la señal de menor frecuencia ($f_B$).

La relación matemática medida de forma directa por el contador es:

$$R = \frac{f_A}{f_B} = \frac{N_A}{N_B}$$

Donde:
*   $f_A$: Frecuencia de la señal más alta ($10\text{ MHz}$).
*   $f_B$: Frecuencia de la señal más baja ($1\text{ MHz}$).
*   $N_A$: Cuentas registradas en el canal A durante el intervalo de medición.
*   $N_B$: Cuentas registradas (o programadas como factor de escala) en el canal B.

#### Evaluación de la Incertidumbre (Guía GUM)

En el modo Ratio $A/B$, el error de la Base de Tiempo interna **se cancela** debido a que ambos canales se sincronizan bajo el mismo reloj base. Las fuentes de incertidumbre remanentes son:

1.  **Incertidumbre Tipo A ($u_A$):** Se evalúa estadísticamente a partir de la desviación estándar muestral de las $n=10$ observaciones individuales.
    $$u_A = \frac{s(R)}{\sqrt{n}}$$

2.  **Incertidumbre Tipo B ($u_B$):** Está dominada puramente por el error de cuantización de $\pm 1$ cuenta en las compuertas de conteo de alta velocidad de los canales. Su valor relativo se modela mediante una distribución rectangular:
    $$u_{B} = R_{\text{media}} \cdot \frac{\Delta N}{\sqrt{3} \cdot N_A}$$
    Dado que buscamos la *mayor resolución posible*, el instrumento minimiza este aporte elevando $N_A$ a su límite instrumental superior.


In [13]:
def incert_B_ratio(R_media, f_A, gate_time=10):
    """
    Calcula la incertidumbre Tipo B para el modo Ratio A/B.
    Para máxima resolución se asume un Gate Time largo (ej. 10s).
    """
    # N_A = total de cuentas estimadas en canal A durante el Gate Time
    N_A = f_A * gate_time
    delta_N = 1  # Error de cuantización instrumental (+/- 1 cuenta)
    
    # Error relativo por cuantización bajo distribución rectangular
    uB_rel = delta_N / (np.sqrt(3) * N_A)
    return R_media * uB_rel

# ==========================================
# CONFIGURACIÓN DE DATOS DEL EXPERIMENTO
# ==========================================

# Muestras
mediciones = [13.123, 12.987, 13.123, 12.987, 13.123, 12.987, 13.123, 12.987, 13.123, 12.987]

f_A_nominal = 10e6    # Canal A: 10 MHz

# Estadísticos iniciales
R_media = np.mean(mediciones)

# Cálculo de incertidumbres estándar
uA_R = incert_A(mediciones)
uB_R = incert_B_ratio(R_media, f_A_nominal, gate_time=10)

# Incertidumbre combinada y expandida
uc_R = np.sqrt(uA_R**2 + uB_R**2)
U_R = k * uc_R

# ==========================================
# IMPRESIÓN DE RESULTADOS
# ==========================================
print(f"--- Análisis Estadístico ---")
print(f"Media Calculada (R): {R_media:.4f}")
print(f"uA (Estadística):    {uA_R:.6f}")
print(f"uB (Instrumental):   {uB_R:.6e}")
print(f"uc (Combinada):      {uc_R:.6f}")
print(f"U  (Expandida k=2):  {U_R:.6f}")

--- Análisis Estadístico ---
Media Calculada (R): 13.0550
uA (Estadística):    0.022667
uB (Instrumental):   7.537308e-08
uc (Combinada):      0.022667
U  (Expandida k=2):  0.045333


### Resultados Finales:

El resultado de la relación de frecuencias se expresa formalmente como:

$$R = 13.055 \pm 0.046$$

*Nota: La incertidumbre expandida reportada ($U = 0.046$) se calculó con un factor de cobertura $k = 2$, un nivel de confianza de aproximadamente el 95%. Se observa que la incertidumbre Tipo A absorbió la casi totalidad de la incertidumbre combinada debido a la inestabilidad de la relación registrada en la muestra, esto puede deberse a  que la compuerta está usando el mismo clock para medición y nosotros lo estamos reinsertando desde el propio instrumento*
